Feature: Lambda Event Filtering for DynamoDB Streams, SQS

merowinger92 opened the pull request:
## Description
This PR enables filtering of Event Filtering for Lambda functions for **DynamoDB Streams and SQS**.
The filtering implements following filtering syntax: https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-syntax


Comparison operator | Example | Rule syntax
-- | -- | --
Null | UserID is null | "UserID": [ null ]
Empty | LastName is empty | "LastName": [""]
Equals | Name is "Alice" | "Name": [ "Alice" ]
And | Location is "New York" and Day is "Monday" | "Location": [ "New York" ], "Day": ["Monday"]
Or | PaymentType is "Credit" or "Debit" | "PaymentType": [ "Credit", "Debit"]
Not | Weather is anything but "Raining" | "Weather": [ { "anything-but": [ "Raining" ] } ]
Numeric (equals) | Price is 100 | "Price": [ { "numeric": [ "=", 100 ] } ]
Numeric (range) | Price is more than 10, and less than or equal to 20 | "Price": [ { "numeric": [ ">", 10, "<=", 20 ] } ]
Exists | ProductName exists | "ProductName": [ { "exists": true } ]
Does not exist | ProductName does not exist | "ProductName": [ { "exists": false } ]
Begins with | Region is in the US | "Region": [ {"prefix": "us-" } ]

### Open Questions
- Batch Sizing while filtering: I didn't make any investigation how the batch size behaves in AWS when using event filters

### Related PR
fixes https://github.com/localstack/localstack/issues/6096
fixes https://github.com/localstack/localstack/issues/5942

dominikschubert commented on localstack/services/awslambda/lambda_utils.py:
```diff
@@ -288,3 +289,110 @@ def generate_lambda_arn(
+    for idx in range(0, len(numeric_filter), 2):
+
+        try:
+            if numeric_filter[idx] == ">" and record_value > float(numeric_filter[idx + 1]):
+                all_comparisons.append(True)
+                continue
+            if numeric_filter[idx] == ">=" and record_value >= float(numeric_filter[idx + 1]):
+                all_comparisons.append(True)
+                continue
+            if numeric_filter[idx] == "=" and record_value == float(numeric_filter[idx + 1]):
+                all_comparisons.append(True)
+                continue
+            if numeric_filter[idx] == "<" and record_value < float(numeric_filter[idx + 1]):
+                all_comparisons.append(True)
+                continue
+            if numeric_filter[idx] == "<=" and record_value <= float(numeric_filter[idx + 1]):
+                all_comparisons.append(True)
+                continue
+            all_comparisons.append(False)
```
nit: could be simplified

dominikschubert replied:
e.g. mapping the operation strings to the corresponding operations. like `>=` to `__ge__()`

dominikschubert replied:
alternatively using the operator module https://docs.python.org/3/library/operator.html might be an option

dominikschubert commented on localstack/services/awslambda/lambda_utils.py:
```diff
@@ -288,3 +289,110 @@ def generate_lambda_arn(
+    for key, value in filter_rule.items():
+        # check if rule exists in event
+        record_value = record.get(key.lower(), record.get(key))
+        if record_value is None:
+            # special case 'exists'
+            if isinstance(value, list) and len(value) > 0:
+                if not value[0].get("exists", True):
+                    filter_results.append(True)
+                    continue
+            filter_results.append(False)
+            continue
+
+        # check if filter rule value is a list (leaf of rule tree) or a dict (rescursively call function)
+        if isinstance(value, list):
+            if len(value) > 0:
+                if isinstance(value[0], (str, int)):
+                    filter_results.append(record_value in value)
+                    continue
+                if isinstance(value[0], dict):
+                    filter_results.append(verify_dict_filter(record_value, value[0]))
+                    continue
+            else:
+                LOG.warn(f"Empty lambda filter: {key}")
+        if isinstance(value, dict):
+            filter_results.append(filter_stream_record(value, record_value))
+            continue
```
please avoid the overuse of continue and/or return statements in favor or restructuring the flow

dominikschubert commented on localstack/services/awslambda/lambda_api.py:
```diff
@@ -236,6 +236,8 @@ def build_mapping_obj(data) -> Dict:
     if data.get("DestinationConfig"):
         mapping["DestinationConfig"] = data.get("DestinationConfig")
 
+    if data.get("FilterCriteria"):
+        mapping["FilterCriteria"] = data.get("FilterCriteria")
```
Please make sure the corresponding CloudFormation model as well :+1:

dominikschubert commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -418,6 +418,129 @@ def verify_failure_received():
+            retry(assert_events, retries=10)
+
+            # putting the same item a second time is a 'MODIFY' request
+            if item_to_put2:
+                dynamodb_client.put_item(TableName=table_name, Item=item_to_put2)
+            else:
+                dynamodb_client.put_item(TableName=table_name, Item=item_to_put1)
+            if calls > 1:
+
+                def assert_events2():
+                    events = get_lambda_log_events(function_name)
+                    assert len(events) == calls
+
+                retry(assert_events2, retries=10)
+            else:
+                retry(assert_events, retries=10)
```
This was a bit hard to understand in combination with the parametrization. Would be amazing if you could add a few more lines here explaining the different scenarios.

dominikschubert reviewed (changes requested):
Some minor nits, mostly in terms of control flow and documentation.

When these are fixed, I'll check the tests against AWS before approving :+1:

dominikschubert commented:
> However I didn't verify the tests against AWS. Is there any trick to do so?

@merowinger92 
To run an integration test against AWS we have an environment variable that we can set (`TEST_TARGET`), i.e.:
```bash
TEST_TARGET="AWS_CLOUD" AWS_PROFILE="<your-aws-profile>" AWS_REGION="us-east-1" python -m pytest ....
```
`<your-aws-profile>` would need to point to an actual AWS profile that you want to test against.

Any test that we've verified to work against AWS in this way, we currently mark with the pytest marker `@pytest.mark.aws_validated`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            ),
+        ],
+    )
+    def test_dynamo_db_insert_event(
```
we would love to make this test compatible with aws, therefore we still need to add policies, and make a few adaptions:

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+        create_lambda_function,
+        lambda_client,
+        dynamodb_client,
+        dynamodb_create_table,
```
please add the following fixtures here:
```
        create_role,
        create_iam_role_with_policy,
        logs_client,
        wait_for_dynamodb_stream_ready,
```

`wait_for_dynamodb_stream_ready` is a new fixture that should ensure that the stream is ready to process. You can add it to `fixtures.py`, something like this would suffice:
```
@pytest.fixture
def wait_for_dynamodb_stream_ready(dynamodbstreams_client):
    def _wait_for_stream_ready(stream_arn: str):
        def is_stream_ready():
            describe_stream_response = dynamodbstreams_client.describe_stream(StreamArn=stream_arn)
            return describe_stream_response["StreamDescription"]["StreamStatus"] == "ENABLED"

        poll_condition(is_stream_ready)

    return _wait_for_stream_ready
```

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+        item_to_put1,
+        item_to_put2,
+    ):
+
```
We need to create role and policy here. The `role_arn` will be used for the lambda in the next step:
```
        role = f"test-lambda-dynamodb-role-{short_uid()}"
        policy_name = f"test-lambda-dynamodb-role-policy-{short_uid()}"
        role_arn = create_iam_role_with_policy(
            RoleName=role,
            PolicyName=policy_name,
            RoleDefinition=lambda_role,
            PolicyDefinition=s3_lambda_permission,
        )
```

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            create_lambda_function(
+                handler_file=TEST_LAMBDA_PYTHON_ECHO,
+                func_name=function_name,
+                runtime=LAMBDA_RUNTIME_PYTHON37,
```
now we need to set the role-arn to the lambda:
`role=role_arn`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            stream_arn = dynamodb_client.update_table(
+                TableName=table_name,
+                StreamSpecification={"StreamEnabled": True, "StreamViewType": "NEW_AND_OLD_IMAGES"},
+            )["TableDescription"]["LatestStreamArn"]
```
after enabling the stream, we should wait until the stream is ready:
`wait_for_dynamodb_stream_ready(stream_arn)`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            event_source_uuid = lambda_client.create_event_source_mapping(
+                FunctionName=function_name,
+                BatchSize=1,
+                StartingPosition="LATEST",
```
for some reason, AWS did not pick up the events, when using `LATEST`. We can change it to `TRIM_HORIZON` here, as we freshly create the table and stream for every test run.

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            dynamodb_client.put_item(TableName=table_name, Item=item_to_put1)
+
+            def assert_single_lambda_call():
+                events = get_lambda_log_events(function_name)
```
in order to use this helper function with AWS, we have to set the `logs_client`: `get_lambda_log_events(function_name, logs_client=logs_client)`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+                events = get_lambda_log_events(function_name)
+                assert len(events) == 1
+
+            retry(assert_single_lambda_call, retries=10)
```
AWS is slow in forwarding the events. We need to increase the retries, I would suggest:
`retry(assert_single_lambda_call, retries=50, sleep=2)`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            if calls > 1:
+
+                def assert_events_called_multiple():
+                    events = get_lambda_log_events(function_name)
```
as above: please set the `logs_client=logs_client` for the helper function in order to retrieve the logs from AWS

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+                    assert len(events) == calls
+
+                # lambda was called a second time, so new records should be found
+                retry(assert_events_called_multiple, retries=10)
```
please increase the `retries` as above

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            (
+                {"id": {"S": "test123"}},
+                None,
+                None,
```
this will set `"FilterCriteria": {"Filters": [{"Pattern": "null"}]}}` and will throw an error when running the test against AWS:
```
An error occurred (InvalidParameterValueException) when calling the CreateEventSourceMapping operation: Invalid filter pattern definition.
```

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+            # Test with OR filter, and two times same entry (aka MODIFY)
+            (
+                {"id": {"S": "test123"}},
+                None,
```
When running the test against AWS, the second insert wasn't recognized as a "modify" event. When I changed the second insert to `{"id": {"S": "test123"},"test": {"S": "modified"}}` it would work. 

There might be some hidden logic in AWS that prevents firing the "modify" when the item did not change at all. Please adapt the test cases accordingly.

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -420,6 +420,139 @@ def verify_failure_received():
+                1,
+            ),
+            # numeric filters
+            (
```
the "numeric filters" and "prefix" did not trigger the lambda at all when running against AWS.

the json would look like this:
```
{
    "Records": [
        {
            "eventID": "f5d9470eb501e25133bb439ddbe7fec3",
            "eventName": "INSERT",
            "eventVersion": "1.1",
            "eventSource": "aws:dynamodb",
            "awsRegion": "us-east-1",
            "dynamodb": {
                "ApproximateCreationDateTime": 1660213902,
                "Keys": {
                    "id": {
                        "S": "test123"
                    }
                },
                "NewImage": {
                    "numericFilter": {
                        "N": "123"
                    },
                    "id": {
                        "S": "test123"
                    }
                },
                "SequenceNumber": "700000000008351815629",
                "SizeBytes": 34,
                "StreamViewType": "NEW_AND_OLD_IMAGES"
            },
            "eventSourceARN": "arn:aws:dynamodb:us-east-1:<accountid>:table/test-table-f7dc1c7b/stream/2022-08-11T10:31:38.042"
        }
    ]
}
```
So I wondered if the filter also needs the `"N"`. 
However, I found a [post on the AWS forum](https://repost.aws/questions/QUxxQDRk5mQ22jR4L3KsTkKQ/dynamo-db-streams-filter-with-nested-fields-not-working), where users where discussing the same issue - the filters were just not working as expected.

Maybe you could find out if this is indeed a problem on AWS side. It would be fine with me, if we leave the test and add the above link for documentation purpose as comment to the test cases.

merowinger92 replied:
I don't have confirmation, but internally entries in DynamoDB are represented as strings (and no number). And tha'ts most likely the reason why numeric filters are not working. 
The result of this is that you can't filter with `numeric`, since there doesn't happen no conversion/cast from string to number.

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
+        )
+
+        response = cfn_client.create_stack(
+            StackName=stack_name, TemplateBody=template_1, Capabilities=["CAPABILITY_IAM"]
```
apparently AWS also needs this capability: `"CAPABILITY_AUTO_EXPAND"`

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
+
+
+def test_lambda_w_dynamodb_event_filter(
+    cfn_client, dynamodb_client, is_stack_created, cleanup_stacks
```
please add `logs_client` as a fixture

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
+        dynamodb_client.put_item(TableName=table_name, Item=item_to_put)
+
+        def assert_single_lambda_call():
+            events = get_lambda_log_events(function_name)
```
similar to the other test: we need to add the `logs_client` so retrieve the logs from AWS `get_lambda_log_events(function_name, logs_client=logs_client)`

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
+            events = get_lambda_log_events(function_name)
+            assert len(events) == 1
+
+        retry(assert_single_lambda_call, retries=10)
```
please increase the `retries` as AWS is slow

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
+
+        # put item the second time: MODIFY and lambda should not be called
+        dynamodb_client.put_item(TableName=table_name, Item=item_to_put)
+        retry(assert_single_lambda_call, retries=10)
```
also: please increase the `retries`

steffyP commented on tests/integration/cloudformation/test_cloudformation_lambda.py:
```diff
@@ -93,3 +95,43 @@ def test_update_lambda_inline_code(
         cleanup_stacks([stack_name])
+
+
+def test_lambda_w_dynamodb_event_filter(
```
with the changes below, the test should work against AWS 👍

steffyP reviewed (changes requested):
Hi @merowinger92,

thank you so much for your contribution! I know a lot of people are eagerly waiting for this feature  🚀 😄 

The implementation looks good, we just need to make sure that the tests also validate against AWS.
Please also re-base with upstream-master to get the latest changes 🙂 

As I understood, you have not yet run your tests against AWS - I tried to add very detailed comments about what is still missing in order to make the tests compatible. 

Please feel free to reach out, if you still have questions - we really want this feature in the codebase and and are here to support you in getting it approved!

One more thing: maybe you could also update the description of the PR to set the expectations straight, as the lambda event-filter for SQS are not implemented with your PR. 🙏

steffyP commented:
Hi @merowinger92
> I validated now the tests aainst AWS and they look like passing in the current form. 

That's great news! 🎉 

> Before compleltely submitting this for review I would like to cover still that `InvalidParameterValueException` case (and maybe investigate if I can also add tests for SQS). 

Perfect, thank you!

> Is there any process for resolving the conversions? Can I do that myself or is that done by someone of you?

We currently don't have a formal process, please feel free to resolve the addressed conversations yourself.

Please don't forget to rebase with master, it seems like the circleci `preflight` step is still failing due to an issue that was already resolved on LocalStack's master.

merowinger92 commented:
Hi @steffyP !
I tried to adopt a SQS test to add a FilterCriteria to get the AWS behaviour with different filters. However I only get `InvalidParamerValueException`. So I won't implement this with this PR.
Furthermore I added also to raise that `InvalidParameterException`, or at least an Exception, since I didn't know the way to corectly raise an Exception so that it get's parsed correctly.

steffyP commented:
Hi @merowinger92,

thanks for the update! 

> I tried to adopt a SQS test to add a FilterCriteria to get the AWS behaviour with different filters. However I only get `InvalidParamerValueException`. So I won't implement this with this PR. 

That's fine, thank you!
On a side note (but I think you already hinted at it with your EDIT note): the [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-sqs) specify that for SQS the body can be any string, but needs to be in a valid JSON if the filter criteria is a JSON-filter - maybe that caused the `InvalidParameterValueException`? 

> Furthermore I added also to raise that `InvalidParameterException`, or at least an Exception, since I didn't know the way to corectly raise an Exception so that it get's parsed correctly.

For all [ASF migrated services](https://localstack.cloud/blog/2022-08-04-parity-explained/), using the ServiceException just as you did is the way to go. However, lambda is not yet migrated, and if you check the [code section here](https://github.com/merowinger92/localstack/blob/mk-ddbstreams-filter/localstack/services/awslambda/lambda_api.py#L1707-L1709), it will only correctly parse exceptions of type `ValueError` to an "error_response".

I will add comments directly to the related code.

steffyP commented on localstack/services/awslambda/lambda_api.py:
```diff
@@ -239,6 +241,14 @@ def build_mapping_obj(data) -> Dict:
+            # AWS raises following Exception when FilterCriteria is not valid:
+            # An error occurred (InvalidParameterValueException) when calling the CreateEventSourceMapping operation:
+            # Invalid filter pattern definition.
+            raise InvalidParameterValueException("Invalid filter pattern definition.")
```
Unfortunately using the `InvalidParameterValueException` will not be parsed to the expected error-response. We need to use `ValueError` like this:
```
raise ValueError(
                INVALID_PARAMETER_VALUE_EXCEPTION,"Invalid filter pattern definition.")
```

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -422,6 +426,174 @@ def verify_failure_received():
+            ),
+        ],
+    )
+    def test_dynamo_db_insert_event(
```
NIT: maybe you could rename this test to something more meaningful? like `test_dynamodb_event_filter`

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -422,6 +426,174 @@ def verify_failure_received():
+                event_source_uuid = lambda_client.create_event_source_mapping(
+                    **event_source_mapping_kwargs
+                )["UUID"]
+            except Exception as ex:
```
I would suggest the following:
- remove this section with try-except
- remove the first parameterized test-values with the `None` filter
- add a new test case that tests invalid filter, eg. `test_dynamodb_invalid_event_filter`

In this new test, you can some invalid filter-pattern, and then use the `pytest.raises` pattern. I already tested the following:

```
@pytest.mark.aws_validated
@pytest.mark.parametrize(
        "filter",
        [
            '[{"eventName": ["INSERT"=123}]',
          'null'
        ]
    )
   def test_dynamodb_invalid_event_filter(self, create_lambda_function,
        lambda_client,
        dynamodb_client,
        dynamodb_create_table,
        lambda_su_role,
        wait_for_dynamodb_stream_ready,
                                           filter):
        # do the setup like in the other test
        event_source_mapping_kwargs.update(
            FilterCriteria={
                "Filters": [
                    {"Pattern": filter},
                ]
            }
        )
        with pytest.raises(Exception) as expected:
            lambda_client.create_event_source_mapping(
                **event_source_mapping_kwargs
            )
        assert expected.typename == "InvalidParameterValueException"
```

steffyP commented on localstack/services/awslambda/lambda_utils.py:
```diff
@@ -299,3 +300,106 @@ def generate_lambda_arn(
+
+def validate_filters(filter: FilterCriteria):
+    for rule in filter["Filters"]:
+        if not json.loads(rule["Pattern"]):
```
Please add a try-except-block here, as some patterns might contain invalid json and will throw a `JSONDecodeError`. In this case we should also return `False`

merowinger92 commented:
Hi @steffyP ! 
I think I now reached a reviewable and maybe mergeable state.

One thing where I didn't spent time to investigate the behaviour is: 
- Batch Sizing while filtering:  how the batch size behaves in AWS when using event filters

I also added that as Open Question to the PR Description.

steffyP commented on localstack/services/awslambda/event_source_listeners/sqs_event_source_listener.py:
```diff
@@ -186,6 +190,20 @@ def delete_messages(result: InvocationResult, func_arn, event, error=None, **kwa
+        if len(event_filter_criterias) > 0:
+            # convert to json for filtering
+            for record in records:
+                record["body"] = json.loads(record["body"])
```
[According to the docs](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-sqs), a non-JSON body would still be a valid message.

> For Amazon SQS, the message body can be any string.

As I understand from the table in the docs: If the incoming message is a plain string, and the filter pattern is a valid json, then the message should be dropped. 
Thus, I think a try-except block here would be sufficient. 

Currently [the exception is caught here and lost](https://github.com/merowinger92/localstack/blob/mk-ddbstreams-filter/localstack/services/awslambda/event_source_listeners/sqs_event_source_listener.py#L77), because we only log a message for the if-condition. 

Maybe you could log the exception in any case there 🙏

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -422,6 +597,225 @@ def verify_failure_received():
+                event_source_uuid = lambda_client.create_event_source_mapping(
+                    **event_source_mapping_kwargs
+                )["UUID"]
+            except ClientError as ex:
+                if filter:
+                    raise ex
+                # json.loads serializes None to "FilterCriteria": {"Filters": [{"Pattern": "null"}]}}
+                assert ex.response["Error"]["Code"] == INVALID_PARAMETER_VALUE_EXCEPTION
+                return
```
NIT: I think we can remove the entire "except" block - there are no input parameters anymore that should throw an expected exception.

steffyP commented on tests/integration/awslambda/test_lambda_integration.py:
```diff
@@ -422,6 +597,225 @@ def verify_failure_received():
+                lambda_client.create_event_source_mapping(**event_source_mapping_kwargs)
+            expected.match(INVALID_PARAMETER_VALUE_EXCEPTION)
+
+        finally:
+            if event_source_uuid:
+                lambda_client.delete_event_source_mapping(UUID=event_source_uuid)
+
```
NIT: We never assign `event_source_uuid` in this test. Maybe you can add the assignment for `create_event_source_mapping` which would be the safe-version in case the expected exception is not thrown.

steffyP reviewed (changes requested):
Hi @merowinger92!

Amazing that you were able to also implement the JSON filter for SQS  🚀 

Thank you so much for this - I can see it was a lot of effort! And also the tests are all validated against AWS, which is great 🎉 

I only found one small issue, I think we still need to address for SQS: it is about messages that do not have a valid JSON body. Those messages are valid [according to the docs](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-sqs). Currently, we assume that the body is always in JSON format, which might not be the case.

The docs also mention "plain string" filters for SQS - but this is something we can address in a future PR (same for the batches).

Besides that, I only marked NITs for cleaner tests.

I would love to merge your implementation as soon as possible, and the JSON filters are already amazing! 😄

merowinger92 commented:
Hi @steffyP !

I addressed your requested changes.
Could you have a look again?

I also rebased to latest master, so that in case you are satisfied, this might be ready to be merged :rocket:.

steffyP reviewed (approved):
Thanks for addressing this so quickly, @merowinger92!
The changes look good, I would only wait for the tests to be finished before merging.

Thanks for your contribution 🚀 👍

steffyP reviewed (changes requested):
Hi @merowinger92,

unfortunately I had some troubles running our extended integration test suite with your branch. 
The main reason seems to be the long-running tests, caused by the high `sleep_before` in the `retry`. 

```
retry(   _check_lambda_logs,
            retries=10,
            sleep_before=30,
        )
```
I am sorry, I overlooked this in the review.  
We need to set the `sleep_before` as low as possible, but can increase the amount of `retries` if required. I already tested it with `sleep_before=2` (which also worked against AWS), but maybe we can even decrease it further. 

I am also looking into some flaky tests, but I am not sure yet if these are connected to your changes.

merowinger92 commented:
@steffyP updated and rebased.

steffyP reviewed (approved):
Thanks for the quick updates and your patience @merowinger92 🙏 

The issue with the long-running tests was resolved, and the flaky tests appear to be unrelated to your changes.

Great work, thank you so much for implementing the event-filtering feature! 🚀 🎉
